File: algorithms/227.Basic_Calculator_II/Basic_Calculator_II.cpp
<<<<<<< HEAD
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int calculate(string s) {
    istringstream in('+' + s + '+');
    long long total = 0, term = 0, n;
    char op;
    while (in >> op) {
        if (op == '+' or op == '-') {
            total += term;
            in >> term;
            term *= 44 - op;
        } else {
            in >> n;
            if (op == '*')
                term *= n;
            else
                term /= n;
        }
    }
    return total;
}
};
 int calculate(string s) {
    int sum = 0; 
    if(s.size() < 1) return sum; 
    int i = 0; 
    int last = 0, last_result = 1;  
    char last_operator = '+'; //remember the last operator
    int sign = 1; 
    while(i < s.size()){
        if(s[i] == ' '){++i; continue;}
        if(s[i] == '+' || s[i] == '-'){
            sum += last_result * sign;
            sign = s[i++] == '+' ? 1 : -1; 
            last_result = 1;
            last_operator = '+'; 
        }
        else if(s[i] == '/' || s[i] == '*'){
            last_operator = s[i];
            ++i; 
        }
        if(isdigit(s[i])){
            last = 0; 
            while(i < s.size() && isdigit(s[i])){
                last =  last * 10 + s[i++] - '0'; 
            }

            if(last_operator == '*') last_result *= last;
            else if(last_operator == '/') last_result /= last;
            else  last_result = last; 
        }
    }
    sum += last_result * sign;
    return sum; 
}
public class Solution {
public int calculate(String s) {
    int len;
    if(s==null || (len = s.length())==0) return 0;
    Stack<Integer> stack = new Stack<Integer>();
    int num = 0;
    char sign = '+';
    for(int i=0;i<len;i++){
        if(Character.isDigit(s.charAt(i))){
            num = num*10+s.charAt(i)-'0';
        }
        if((!Character.isDigit(s.charAt(i)) &&' '!=s.charAt(i)) || i==len-1){
            if(sign=='-'){
                stack.push(-num);
            }
            if(sign=='+'){
                stack.push(num);
            }
            if(sign=='*'){
                stack.push(stack.pop()*num);
            }
            if(sign=='/'){
                stack.push(stack.pop()/num);
            }
            sign = s.charAt(i);
            num = 0;
        }
    }

    int re = 0;
    for(int i:stack){
        re += i;
    }
    return re;
}
int main() {
    return 0;
}

=======
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int calculate(string s) {
    istringstream in('+' + s + '+');
    long long total = 0, term = 0, n;
    char op;
    while (in >> op) {
        if (op == '+' or op == '-') {
            total += term;
            in >> term;
            term *= 44 - op;
        } else {
            in >> n;
            if (op == '*')
                term *= n;
            else
                term /= n;
        }
    }
    return total;
}
};
 int calculate(string s) {
    int sum = 0; 
    if(s.size() < 1) return sum; 
    int i = 0; 
    int last = 0, last_result = 1;  
    char last_operator = '+'; //remember the last operator
    int sign = 1; 
    while(i < s.size()){
        if(s[i] == ' '){++i; continue;}
        if(s[i] == '+' || s[i] == '-'){
            sum += last_result * sign;
            sign = s[i++] == '+' ? 1 : -1; 
            last_result = 1;
            last_operator = '+'; 
        }
        else if(s[i] == '/' || s[i] == '*'){
            last_operator = s[i];
            ++i; 
        }
        if(isdigit(s[i])){
            last = 0; 
            while(i < s.size() && isdigit(s[i])){
                last =  last * 10 + s[i++] - '0'; 
            }

            if(last_operator == '*') last_result *= last;
            else if(last_operator == '/') last_result /= last;
            else  last_result = last; 
        }
    }
    sum += last_result * sign;
    return sum; 
}
public class Solution {
public int calculate(String s) {
    int len;
    if(s==null || (len = s.length())==0) return 0;
    Stack<Integer> stack = new Stack<Integer>();
    int num = 0;
    char sign = '+';
    for(int i=0;i<len;i++){
        if(Character.isDigit(s.charAt(i))){
            num = num*10+s.charAt(i)-'0';
        }
        if((!Character.isDigit(s.charAt(i)) &&' '!=s.charAt(i)) || i==len-1){
            if(sign=='-'){
                stack.push(-num);
            }
            if(sign=='+'){
                stack.push(num);
            }
            if(sign=='*'){
                stack.push(stack.pop()*num);
            }
            if(sign=='/'){
                stack.push(stack.pop()/num);
            }
            sign = s.charAt(i);
            num = 0;
        }
    }

    int re = 0;
    for(int i:stack){
        re += i;
    }
    return re;
}
int main() {
    return 0;
}

>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/227.Basic_Calculator_II/Basic_Calculator_II.py
def calculate(s):
    num, op, stack = 0, '+', [0]
    ops = {'+':lambda x, y: y, '-':lambda x, y: -y, '*':lambda x, y: x*y, '/':lambda x, y: (int)(float(x)/float(y))}
    for i, c in enumerate(s):
        if c.isdigit():
            num = num * 10 + int(c)
        if not c.isdigit() and c != ' ' or i == len(s)-1:
            prev = 0 if op in '+-' else stack.pop()
            stack.append(ops[op](prev, num))
            num, op = 0, c
    return sum(stack)

class Solution:
    # @param {string} s
    # @return {integer}
    def calculate(self, s):
        operands, operators = [], []
        operand = ""
        for i in reversed(xrange(len(s))):
            if s[i].isdigit():
                operand += s[i]
                if i == 0  or not s[i-1].isdigit():
                    operands.append(int(operand[::-1]))
                    operand = ""
            elif s[i] == ')' or s[i] == '*' or s[i] == '/':
                operators.append(s[i])
            elif s[i] == '+' or s[i] == '-':
                while operators and \
                      (operators[-1] == '*' or operators[-1] == '/'):
                    self.compute(operands, operators)
                operators.append(s[i])
            elif s[i] == '(':
                while operators[-1] != ')':
                    self.compute(operands, operators)
                operators.pop()
                
        while operators:
            self.compute(operands, operators)
            
        return operands[-1]

    def compute(self, operands, operators):
        left, right = operands.pop(), operands.pop()
        op = operators.pop()
        if op == '+':
            operands.append(left + right)
        elif op == '-':
            operands.append(left - right)
        elif op == '*':
            operands.append(left * right)
        elif op == '/':
            operands.append(left / right)


def calculate(s):
    l=list(s)
    if len(l) == 0: return 0;
    num=0
    sign='+'
    stack = []
    for idx,i in enumerate(l):
        if i.isdigit():
            num=num*10+int(i)
        # when the signs occurs,time to calculate your number
        if (i in ('+','-','*','/')) or idx==len(l)-1:
            if sign=='+':
                stack.append(num)
            if sign=='*':
                stack.append(int(stack.pop()) * num)
            if sign=='-':
                stack.append(-num)
            if sign=='/':
                top = int(stack.pop())
                if top >= 0:
                    stack.append( top / num)
                else:
                    stack.append(-( abs(top) / num  ))
            sign = i 
            num=0
    return sum(stack)
